doc/source/scipeps/1-Optimizer.rst

```diff
@@ -0,0 +1,981 @@
+	* If this had been present when development of your library began, how would have it influenced your library?
+	* Libraries: sklearn, skimage, cvxpy, daskml, PyTorch, theano, Chainer, neon, Thinc
+
+.. outline
+
+   * Abstract
+   * Introduction
+       * Here's what minimization does...
+           * It minimizes a function
+           * These are -or should be- fairly independent -- functions and optimizers are not tied together.
+       * Point to users of...
+           * Minimization in general
+           * scipy.optimize.minimize (many users, do a github search)
+   * Proposed solution
+       * Classes (idea: `Function` and `Optimizer` class)
+           * `Optimizer` - takes care of minimization and stepping
+           * `Function` - takes care of evaluating function, gradient, and hessian.
+       * Goals:
+            * enhancing ease of use for ``minimize``
+            * API cleaning and maintainability of ``minimize``
+            * preserving backwards compatibility
+            * exposing a new API to easily create optimizers
+       * Example
+   * Goals
+       * enhancing ease of use for ``minimize``
+           * Have to explain why minimize isn't a standard interface.
+       * preserving backwards compatibility
+       * exposing a new API to easily create optimizers
+           * Provide standard interface for operation
+           * Provide class features
+       * cleaning the existing API
+          * addition of new features to minimizers leads to lengthy functions and lots of duplicate code.
+          * minimize is trying to be a class
+          * function arg is trying to be a class
+          * there is no separation of concerns between function and minimizer
+          * scipy.optimize.minimize is a black box (have to explain why)
+   * Existing work
+       * Class defs: PyTorch, skopt
+       * Functional class wrapper around minimize: statsmodels, astropy, scikits.fitting
+       * Functional defs: sklearn, daskml, skimage
+       * Other:
+         * scikit.optimization (class based, no webpage (download from PyPI)).
+   * Concerns
+       * `minimize` is supposed to implement a unified interface
+          (rewrite from fmin, fmin_bfgs, etc => mininimize)
+       * Why not apply to other solvers in `show_options`? `root`,
+         `minimize_scalar`, `linprog`?
+   * Open bugs
+   * Implementation
+       * List functions, attributes in more depth
+       * Scope
+       * Existing code
+           * How would it work with C/Fortran optimizers?
+           * What interface are we proposing? See proposed code below
+       * Speed
+
++----------+------------------------------------------------------------------------+
+| PEP:     |  1                                                                     |
++----------+------------------------------------------------------------------------+
+| Title:   | Introduction of Optimizer and Function classes for scalar minimisation |
```

Comment:
`Optimizer` makes sense as a name; it would be good to replace `Function` with something more specific

Reply:
The #8328 is using `ScalarFunction` and `VectorFunction`, but not making them public. Those classes are very similar to what is proposed for this approach, but there are some technical differences in current implementation which would have to be ironed out. For example, tracking number of function evaluations, how some memoization is carried out, etc.